README.md
# [Slider](https://github.com/fengyuanchen/slider)

A simple jQuery slide plugin.

- [Demo](https://fengyuanchen.github.io/slider)



## Table of contents

- [Main](#main)
- [Getting started](#getting-started)
- [Options](#options)
- [Methods](#methods)
- [No conflict](#no-conflict)
- [Browser support](#browser-support)
- [Versioning](#versioning)
- [License](#license)



## Main

```
dist/
├── slider.js     (10 KB)
└── slider.min.js ( 5 KB)
```



## Getting started

### Quick start

Three quick start options are available:

- [Download the latest release](https://github.com/fengyuanchen/slider/archive/master.zip).
- Clone the repository: `git clone https://github.com/fengyuanchen/slider.git`.
- Install with [NPM](http://npmjs.org): `npm install @fengyuanchen/slider`.


### Installation

Include files:

```html
<script src="/path/to/jquery.js"></script><!-- jQuery is required -->
<script src="/path/to/slider.js"></script>
```


### Usage

Initialize with `$.fn.slider` method.

```js
$('#slider').slider({
  activeClass: 'slider-active', // for slide nav
  autoPlay: true,
  contentClass: 'slider-content',
  disableClass: 'slider-disabled', // for slide "prev" and "next" button
  duration: 5000,
  effect: 'fade',
  easing: 'swing', // support "swing" and "linear", powered by jQuery
  navClass: 'slider-nav',
  nextClass: 'slider-next',
  prevClass: 'slider-prev',
  speed: 1000,
  trigger: 'click' // for slide nav
});
```

[⬆ Back to top](#table-of-contents)



## Options

```js
// Set slider options
$().slider(options);

// Set the global default options
$.fn.slider.setDefaults(options);
```

### activeClass

- Type: `String`
- Default: `'slider-active'`

An active state class for nav items.


### autoPlay

- Type: `Boolean`
- Default: `true`

Auto play the slide after initialized.


### contentClass

- Type: `String`
- Default: `'slider-content'`

The content container.


### disableClass

- Type: `String`
- Default: `'slider-disabled'`

A disable state class for previous and next button.


### duration

- Type: `Number`
- Default: `5000`

The duration milliseconds of each slide item.


### effect

- Type: `String`
- Options: `'fade'`, `'scrollX'`, `'scrollY'`
- Default: `'fade'`


### easing

- Type: `String`
- Options: `'swing'`, `'linear'`
- Default: `'swing'`

Powered by jQuery.


### navClass

- Type: `String`
- Default: `'slider-nav'`

The navigation container.


### prevClass

- Type: `String`
- Default: `'slider-prev'`

The previous button.


### nextClass

- Type: `String`
- Default: `'slider-next'`

The next button.


### speed

- Type: `Number`
- Default: `1000`

The speed of the animation.


### trigger

- Type: `String`
- Default: `'click'`

The event for triggering slide.


[⬆ Back to top](#table-of-contents)



## Methods

### render()

Render the slider.


### start()

Start the slider.

```js
$().slider({
  autoPlay: false
}).slider('start');
```


### stop()

stop the slider.


### enable()

Enable the slider.


### disable()

Disable the slider.


### destroy()

Destroy the slider.


[⬆ Back to top](#table-of-contents)



## No conflict

If you have to use other plugin with the same namespace, just call the `$.fn.slider.noConflict` method to revert to it.

```html
<script src="other-plugin.js"></script>
<script src="slider.js"></script>
<script>
  $.fn.slider.noConflict();
  // Code that uses other plugin's "$().slider" can follow here.
</script>
```



## Browser Support

- Chrome (latest)
- Firefox (latest)
- Safari (latest)
- Opera (latest)
- Edge (latest)
- Internet Explorer 8+

As a jQuery plugin, you also need to see the [jQuery Browser Support](http://jquery.com/browser-support/).



## Versioning

Maintained under the [Semantic Versioning guidelines](http://semver.org/).



## License

[MIT](http://opensource.org/licenses/MIT) © [Fengyuan Chen](http://chenfengyuan.com)


[⬆ Back to top](#table-of-contents)


CHANGELOG.md
# Changelog

### 0.1.1 (May 27, 2016)

- Simplify code.


### 0.1.0 (May 27, 2016)

- Added a new method: `destroy`.
- Renamed the `autoplay` option to `autoPlay`.
- Changed the available values of the `effect` option:
  - Rename `scrollx` to `scrollX`
  - Rename `scrolly` to `scrollY`
- Improved the cycle of the fade effect.
- Stop sliding when mouse over the slider container.
- Optimized the code style.


### 0.0.2 (Jun 10, 2015)

- Fixed a bug.


### 0.0.1 (Mar 12, 2014)

- Supported options: "activeClass", "autoplay", "contentClass", "disableClass", "duration", "effect", "easing", "navClass", "nextClass", "prevClass", "speed" and "trigger".
- Supported methods: "render", "start", "stop", "enable" and "disable".


gulpfile.js
'use strict';

var gulp = require('gulp');
var plugins = require('gulp-load-plugins')();
var pkg = require('./package');
var now = new Date();
var scripts = {
  all: [
    'gulpfile.js',
    'src/slider.js',
    'docs/js/main.js'
  ],
  docs: 'docs/js',
  src: 'src/slider.js',
  dest: 'dist'
};
var replacement = {
  regexp: /@\w+/g,
  filter: function (placeholder) {
    switch (placeholder) {
      case '@VERSION':
        placeholder = pkg.version;
        break;

      case '@YEAR':
        placeholder = now.getFullYear();
        break;

      case '@DATE':
        placeholder = now.toISOString();
        break;
    }

    return placeholder;
  }
};

gulp.task('jscopy', function () {
  return gulp.src(scripts.src)
  .pipe(gulp.dest(scripts.docs))
  .pipe(gulp.dest(scripts.dest));
});

gulp.task('jshint', function () {
  return gulp.src(scripts.all)
    .pipe(plugins.jshint())
    .pipe(plugins.jshint.reporter('default'));
});

gulp.task('jscs', function () {
  return gulp.src(scripts.all)
    .pipe(plugins.jscs())
    .pipe(plugins.jscs.reporter());
});

gulp.task('js', ['jshint', 'jscs'], function () {
  return gulp.src(scripts.src)
    .pipe(plugins.replace(replacement.regexp, replacement.filter))
    .pipe(gulp.dest(scripts.dest))
    .pipe(gulp.dest(scripts.docs))
    .pipe(plugins.rename({
      suffix: '.min'
    }))
    .pipe(plugins.uglify({
      preserveComments: 'license'
    }))
    .pipe(gulp.dest(scripts.dest));
});

gulp.task('htmlcomb', function () {
  return gulp.src('docs/*.html')
    .pipe(plugins.htmlcomb())
    .pipe(gulp.dest('docs'));
});

gulp.task('docs', ['js'], function () {
  return gulp.src('docs/**')
    .pipe(gulp.dest('_gh_pages'));
});

gulp.task('release', ['docs'], function () {
  return gulp.src('dist/*.{js,css}')
    .pipe(gulp.dest('_releases/' + pkg.version));
});

gulp.task('watch', function () {
  gulp.watch(scripts.src, ['jscopy']);
  gulp.watch('docs/**', ['docs']);
});

gulp.task('default', ['watch']);


Gruntfile.js
module.exports = function(grunt) {

    "use strict";

    grunt.initConfig({
        pkg: grunt.file.readJSON("package.json"),
        banner: "/*!\n" +
                " * Slider v<%= pkg.version %>\n" +
                " * <%= pkg.homepage %>\n" +
                " *\n" +
                " * Copyright <%= grunt.template.today('yyyy') %> <%= pkg.author.name %>\n" +
                " * Released under the <%= pkg.license.type %> license\n" +
                " */\n",
        clean: {
            dist: ["dist/"],
            build: ["build/<%= pkg.version %>.<%= grunt.template.today('yyyymmdd') %>"],
            release: ["release/<%= pkg.version %>"],
            docs: ["../fengyuanchen.github.io/<%= pkg.name %>/"]
        },
        jshint: {
            options: {
                jshintrc: ".jshintrc"
            },
            files: ["Gruntfile.js", "src/<%= pkg.name %>.js"]
        },
        uglify: {
            dist: {
                src: "src/<%= pkg.name %>.js",
                dest:  "dist/<%= pkg.name %>.min.js"
            }
        },
        usebanner: {
            options: {
                position: "top",
                banner: "<%= banner %>"
            },
            files: ["dist/*.js"]
        },
        copy: {
            dist: {
                expand: true,
                cwd: "src/",
                src: "*.js",
                dest: "dist/",
                filter: "isFile"
            },
            build: {
                expand: true,
                cwd: "dist/",
                src: "*.js",
                dest: "build/<%= pkg.version %>.<%= grunt.template.today('yyyymmdd') %>/",
                filter: "isFile"
            },
            release: {
                expand: true,
                cwd: "dist/",
                src: "*.js",
                dest: "release/<%= pkg.version %>/",
                filter: "isFile"
            },
            sync: {
                expand: true,
                cwd: "dist/",
                src: "**",
                dest: "../fengyuanchen.github.io/dist/",
                filter: "isFile"
            },
            i18n: {
                expand: true,
                cwd: "i18n/",
                src: "**",
                dest: "../fengyuanchen.github.io/i18n/",
                filter: "isFile"
            },
            docs: {
                expand: true,
                cwd: "docs/",
                src: "**",
                dest: "../fengyuanchen.github.io/<%= pkg.name %>/",
                filter: "isFile"
            }
        },
        watch: {
            files: ["src/*.js"],
            tasks: "default"
        }
    });

    // Loading dependencies
    require("load-grunt-tasks")(grunt);

    grunt.registerTask("release", ["clean:release", "copy:release"]);
    grunt.registerTask("docs", ["clean:docs", "copy:sync", "copy:i18n", "copy:docs"]);
    grunt.registerTask("default", ["clean:dist", "clean:build", "jshint", "uglify", "copy:dist", "usebanner", "copy:build"]);
};


js/slider.js
/*!
 * Slider v0.1.1
 * https://github.com/fengyuanchen/slider
 *
 * Copyright (c) 2014-2016 Fengyuan Chen
 * Released under the MIT license
 *
 * Date: 2016-05-27T09:58:41.538Z
 */

(function (factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as anonymous module.
    define('slider', ['jquery'], factory);
  } else if (typeof exports === 'object') {
    // Node / CommonJS
    factory(require('jquery'));
  } else {
    // Browser globals.
    factory(jQuery);
  }
})(function ($) {

  'use strict';

  var NAMESPACE = 'slider';

  function isNumber(n) {
    return typeof n === 'number';
  }

  function isString(n) {
    return typeof n === 'string';
  }

  function isUndefined(n) {
    return typeof n === 'undefined';
  }

  function toArray(obj, offset) {
    var args = [];

    // This is necessary for IE8
    if (isNumber(offset)) {
      args.push(offset);
    }

    return args.slice.apply(obj, args);
  }

  function Slider(element, options) {
    var self = this;

    self.$element = $(element);
    self.options = $.extend({}, Slider.DEFAULTS, $.isPlainObject(options) ? options : {});
    self.init();
  }

  Slider.prototype = {
    constructor: Slider,

    init: function () {
      var self = this;
      var $this = self.$element;
      var options = self.options;
      var styles = {
            overflow: 'hidden'
          };

      self.$content = $this.find('.' + options.contentClass);
      self.$items = self.$content.children();

      self.$nav = $this.find('.' + options.navClass);
      self.$btns = self.$nav.children();

      self.$prev = $this.find('.' + options.prevClass);
      self.$next = $this.find('.' + options.nextClass);

      if ($this.css('position') === 'static') {
        styles.position = 'relative';
      }

      $this.css(styles);

      self.index = 0;
      self.length = 1;

      if (self.$items.length > 1) {
        self.render();
      }
    },

    render: function () {
      var self = this;
      var $this = self.$element;
      var options = self.options;
      var firstItem = self.$items.first();

      firstItem.removeAttr('style');
      self.itemHeight = firstItem.height();
      self.itemWidth = firstItem.width();
      self.itemLength = self.$items.length;

      self.width = $this.width();
      self.height = $this.height();

      if (options.effect.toLowerCase() === 'scrollx') {
        self.scrollX();
      } else if (options.effect.toLowerCase() === 'scrolly') {
        self.scrollY();
      } else {
        self.fade();
      }

      self.firstIndexOflastView = self.itemLength - (self.itemLength % self.length || self.length);
      self.enable();

      if (options.autoPlay || options.autoplay) {
        self.start();
      }
    },

    rerender: function () {
      var self = this;
      var $this = self.$element;

      if ($this.width() !== self.width || $this.height() !== self.height) {
        self.stop();
        self.render();
      }
    },

    fade: function () {
      var self = this;

      if (self.$content.css('position') === 'static') {
        self.$content.css('position', 'relative');
      }

      self.$items.css({
        position: 'absolute',
        top: 0,
        left: 0,
        display: 'none',
        width: self.itemWidth,
        height: self.itemHeight
      }).first().show();

      self.sliding = function () {
        var speed = self.options.speed;
        var easing = self.options.easing;

        self.$items.stop(true).fadeOut(speed, easing).eq(self.index).fadeIn(speed, easing);
      };
    },

    scrollX: function () {
      var self = this;

      self.length = Math.floor(self.width / self.itemWidth);

      self.$items.css({
        float: 'left',
        width: self.itemWidth,
        height: self.itemHeight
      });

      self.$content.css({
        position: 'absolute',
        left: 0,
        top: 0,
        width: self.itemWidth * (self.itemLength + 1),
        height: self.itemHeight
      });

      self.sliding = function () {
        self.$content.stop(true).animate({
          left: -self.itemWidth * self.index
        }, self.options.speed, self.options.easing);
      };
    },

    scrollY: function () {
      var self = this;

      self.length = Math.floor(self.height / self.itemHeight);

      self.$items.css({
        width: self.itemWidth,
        height: self.itemHeight
      });

      self.$content.css({
        height: self.itemHeight * (self.itemLength + 1),
        position: 'absolute',
        top: 0,
        left: 0
      });

      self.sliding = function () {
        self.$content.stop(true).animate({
          top: -self.itemHeight * self.index
        }, self.options.speed, self.options.easing);
      };
    },

    enable: function () {
      var self = this;

      if (self.active) {
        return;
      }

      self.active = true;

      self.$element.on({
        mouseover: $.proxy(self.stop, self),
        mouseout: $.proxy(self.start, self)
      });

      self.$btns.on(self.options.trigger, function () {
        self.index = $(self).index();
        self.slide();
      });

      self.$prev.on('click', $.proxy(self.prev, self));
      self.$next.on('click', $.proxy(self.next, self));

      $(window).on('resize', $.proxy(self.resize, self));
    },

    disable: function () {
      var self = this;

      if (!self.active) {
        return;
      }

      self.active = false;
      self.stop();

      self.$element.off({
        mouseover: self.stop,
        mouseout: self.start
      });

      self.$btns.off(self.options.trigger);

      self.$prev.off('click', self.prev);
      self.$next.off('click', self.next);

      $(window).off('resize', self.resize);
    },

    resize: function () {
      var self = this;

      if (self.resizing) {
        clearTimeout(self.resizing);
        self.resizing = null;
      }

      self.resizing = setTimeout($.proxy(self.rerender, self), 200);
    },

    start: function () {
      var self = this;

      if (self.active && !self.autoSlided) {
        self.autoSlided = true;
        self.autoSliding = setInterval($.proxy(self.next, self), self.options.duration);
      }
    },

    stop: function () {
      var self = this;

      if (self.autoSlided) {
        self.autoSlided = false;
        clearInterval(self.autoSliding);
      }
    },

    prev: function () {
      var self = this;
      var prev = self.index - self.length;
      var index;

      if (prev < 0) {
        if (self.options.effect === 'fade') {
          index = self.firstIndexOflastView;
        } else {
          index = 0;
        }
      } else {
        index = prev;
      }

      self.index = index;
      self.slide();
    },

    next: function () {
      var self = this;
      var next = self.index + self.length;
      var index;

      if (next <= self.firstIndexOflastView) {
        index = next;
      } else if (self.autoSlided || self.options.effect === 'fade') {
        index = 0;
      } else {
        index = self.firstIndexOflastView;
      }

      self.index = index;
      self.slide();
    },

    prevable: function () {
      var self = this;
      var prevable = self.index > 0;

      self.$prev.toggleClass(self.options.disableClass, !prevable);

      return prevable;
    },

    nextable: function () {
      var self = this;
      var nextable = self.index < self.firstIndexOflastView;

      self.$next.toggleClass(self.options.disableClass, !nextable);

      return nextable;
    },

    slide: function () {
      var self = this;
      var activeClass = self.options.activeClass;
      var $target = self.$btns.eq(self.index);

      if (!$target.hasClass(activeClass)) {
        $target.addClass(activeClass).siblings().removeClass(activeClass);

        if (self.options.effect !== 'fade') {
          self.prevable();
          self.nextable();
        }

        self.sliding();
      }
    },

    sliding: $.noop,

    destroy: function () {
      var self = this;

      self.disable();
      self.$element.removeData(NAMESPACE);
    }
  };

  // Default settings
  Slider.DEFAULTS = {
    activeClass: 'slider-active', // for slide nav
    autoPlay: true,
    contentClass: 'slider-content',
    disableClass: 'slider-disabled', // for slide 'prev btn' and 'next btn'
    duration: 5000,
    effect: 'fade',
    easing: 'swing', // support 'swing' and 'linear', powered by jquery
    navClass: 'slider-nav',
    nextClass: 'slider-next',
    prevClass: 'slider-prev',
    speed: 1000,
    trigger: 'click' // for slide nav
  };

  // Set default settings
  Slider.setDefaults = function (options) {
    $.extend(Slider.DEFAULTS, options);
  };

  // Save the other slider
  Slider.other = $.fn.slider;

  // Register as jQuery plugin
  $.fn.slider = function (options) {
    var args = toArray(arguments, 1);
    var result;

    this.each(function () {
      var $this = $(this);
      var data = $this.data(NAMESPACE);
      var fn;

      if (!data) {
        if (/destroy/.test(options)) {
          return;
        }

        $this.data(NAMESPACE, (data = new Slider(this, options)));
      }

      if (isString(options) && $.isFunction(fn = data[options])) {
        result = fn.apply(data, args);
      }
    });

    return isUndefined(result) ? this : result;
  };

  $.fn.slider.Constructor = Slider;
  $.fn.slider.setDefaults = Slider.setDefaults;

  // No conflict
  $.fn.slider.noConflict = function () {
    $.fn.slider = Slider.other;
    return this;
  };

  $(function () {
    $('[data-toggle="slider"]').slider();
  });
});
